job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java

```diff
@@ -106,10 +111,24 @@ public void sendMessage(
             throw new RuntimeException(e);
         }
 
+        final String stagingQueueName;
+        if (CAF_WMP_ENABLED) {
+            // TODO
+            // 1) partitionId is tenant-rory, should we try to parse tenantId from it, or just use partitionId. Does a
+            // partitionId always have a tenantId? Maybe we should only call the message router if partitionId.startsWith("tenant-")?
+            // 2) This reroutes messages to whatever worker the targetQueue is for, do we need something like the
+            // action.applyMessagePrioritization field in the workflows to enable/disable rerouting for a worker?
+            // 3) There is no workflow name attached to the staging queue here. Do we need it? How do we get it?
+            MESSAGE_ROUTER_SINGLETON.init();
+            stagingQueueName = MESSAGE_ROUTER_SINGLETON.route(targetQueue, partitionId);
```

Comment:
@andyreidz @dermot-hardy I've made an attempt here at updating the Job Service Scheduled Executor to allow it to reroute messages. This is based on my updates to the MessageRouterSingleton [here](https://github.com/WorkerFramework/worker-message-prioritization/pull/12/files).

I had 3 questions that would be good to chat about (in the comments above). 

For (3), some additional info:

Looking at the jobs in [darwin-job-service](https://github.houston.softwaregrp.net/Verity/darwin-job-service), I think these are the first workers involved in each workflow (taskPipe). I believe these are the workers the Job Service Scheduled Executor will route to for each job:

Batch Ingestion: Batch Ingestion Worker
Batch Enrichment: Data Enrichment Worker
Disposition: Elastic Query Worker
Add Target References: Elastic Worker
Tagging: Elastic Query Worker
Legacy Tagging: Elastic Query Worker
Data Analysis: Elastic Query Worker
AJP Worker: Data Enrichment Worker
AJP Worker Staging Complete: AJP Worker
Random Sampling: Sampling Worker
Clear Failures: Elastic Query Worker
Protection: Elastic Query Worker or Elastic Worker
Workspace Archive Cleanup: App Resources Worker
Released Hold Cleanup: Elastic Query Worker
Repository Initialization: App Resources Worker
Mark Actionable Items: Elastic Query Worker
Grammar Worker: Grammar Worker

So right now, we have no way of controlling whether the Job Service Scheduled Executor will create staging queues for these workers, it always will. 

I think we might need something similar to the [action.applyMessagePrioritization](https://github.com/CAFDataProcessing/worker-workflow/pull/107/files#diff-4e4d4e9550800b6e65e256df7a6790ba18e18ce24f011afd660c6f36b1b1401bR158) flag that we have in the workflow-worker.

Just not sure what would be the equivalent here,  an extra field like this we would add to each job and we would try to parse out here?

```
taskMessageParams": {
    "customdata:applyMessagePrioritization": true
```

This is an example of the taskData that is sent here in the Batch Ingestion Job:

```
{
    "batchType": "IngestionBatchWorkerPlugin",
    "batchDefinition": "rory3/ingestion-1719-187-350-109",
    "targetPipe": "appresources-worker-in",
    "taskMessageParams": {
        "customdata:tenantId": "rory3",
        "customdata:action": "ingestion",
        "graaljs:processFailures.js": "load('classpath:scripts/processFailures.js');",
        "graaljs:setResponse.js": "let getExtractMaxMatches;\n\nfunction onProcessTask(e) {\n  let response = e.task.getResponse();\n  let outputQueue = e.task.getCustomData('sr:outputQueue');\n  response.getSuccessQueue().set(outputQueue);\n  response.getFailureQueue().set(outputQueue);\n  const responseCustomData = response.getCustomData();\n  responseCustomData.putAll(JSON.parse(e.task.getCustomData('sr:customData')));\n\n  const extractMaxMatchesOverride = e.task.getCustomData('sr:extractMaxMatches');\n  getExtractMaxMatches = extractMaxMatchesOverride\n    ? () => extractMaxMatchesOverride\n    : () => {\n      const operationModeParameter = responseCustomData.get('TASK_SETTING_EE.OPERATIONMODE');\n      return (operationModeParameter === 'EXTRACT') ? '100' : '1';\n    }\n\n  setExtractMaxMatches(responseCustomData);\n}\n\nfunction onAfterProcessTask(e) {\n  onError = null;\n  setIngestionWorkflowParams(e.rootDocument);\n}\n\nfunction onError(e) {\n  setIngestionWorkflowParams(e.rootDocument);\n}\n\nfunction setIngestionWorkflowParams(rootDocument) {\n  const customData = rootDocument.getTask().getResponse().getCustomData();\n  const repositoryExtractGrammarValuesField = rootDocument.getField('REPOSITORY_EXTRACT_GRAMMAR_VALUES');\n  const repositoryStoreContentField = rootDocument.getField('REPOSITORY_STORE_CONTENT');\n\n  const operationModeParameter = customData.get('TASK_SETTING_EE.OPERATIONMODE');\n  if (operationModeParameter !== 'EXTRACT' && isFlagSet(repositoryExtractGrammarValuesField)) {\n    customData.put('TASK_SETTING_EE.OPERATIONMODE', 'EXTRACT');\n    setExtractMaxMatches(customData);\n  }\n\n  const storeContentParameter = customData.get('TASK_SETTING_STORE_CONTENT');\n  if (storeContentParameter === 'false' && isFlagSet(repositoryStoreContentField)) {\n    customData.put('TASK_SETTING_STORE_CONTENT', 'true');\n  }\n\n  repositoryExtractGrammarValuesField.reset();\n  repositoryStoreContentField.reset();\n\n  function isFlagSet(field) {\n    return field.hasValues()\n      ? field.getStringValues().get(0) === 'true'\n      : false;\n  }\n}\n\nfunction setExtractMaxMatches(responseCustomData) {\n  responseCustomData.put('TASK_SETTING_ENTITY_EXTRACT_MAX_MATCHES', getExtractMaxMatches());\n}",
        "customdata:sr:outputQueue": "dataprocessing-workflow-in",
        "customdata:sr:customData": "{\"tenantId\":\"rory3\",\"workflowName\":\"ingestion-workflow\",\"settingsServiceLastUpdateTimeMillis\":null,\"TASK_SETTING_EE.OPERATIONMODE\":\"EXTRACT\",\"TASK_SETTING_EE.GRAMMARMAP\":null,\"TASK_SETTING_ENTITY_EXTRACT_LIMITER\":\"CountPerEntityLimiter\",\"TASK_SETTING_STORE_CONTENT\":\"true\",\"extraFailuresSubfieldKey0\":\"AJP_JOB_RUN_ID\",\"extraFailuresSubfieldValue0\":\"187\",\"extraFailuresSubfieldKey1\":\"AJP_WORK_UNIT_ID\",\"extraFailuresSubfieldValue1\":\"350\"}",
        "graaljs:recordProcessingTimes.js": "load('classpath:scripts/recordProcessingTimes.js');",
        "graaljs:complyWithIngestionStatus.js": "load('classpath:scripts/complyWithIngestionStatus.js');"
    }
}
```

Reply:
- On (1), the use of that pattern or how the tenant is extracted from it should not be hardcoded.  Perhaps a regular expression or something along those lines may be acceptable?
- On (2), would it work if you had a new environment variable with a comma-separated list of queues to apply prioritization to?  Or instead of a comma-separated list, a regular expression again, so that you could use wildcards in it (i.e. something like `^(?>dataprocessing-.*-in|worker-grammar-in)$`)?
- On (3), I suppose the job type would be the obvious thing to use for this but I suspect you don't have it at this point.  I'd say do without it.

Reply:
Thanks, I've made those updates now.